From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 31 May 2021 11:30:38 +0200
Subject: src: add impl for EVP_PKEY_get0

This adds the missing implementation for EVP_PKEY_get0, which Node.js
uses and which is present in OpenSSL but not BoringSSL.

This will be upstreamed either fully or as a shim.

diff --git a/crypto/evp/evp.c b/crypto/evp/evp.c
index 653d6573fdf1c542010698e9f6d1cf8170ea2f0c..738cec60a9fed4bfad56c7b2b75c44eb225abfcf 100644
--- a/crypto/evp/evp.c
+++ b/crypto/evp/evp.c
@@ -224,6 +224,10 @@ int EVP_PKEY_type(int nid) {
   return meth->pkey_id;
 }
 
+void *EVP_PKEY_get0(const EVP_PKEY *pkey) {
+  return pkey->pkey.ptr;
+}
+
 int EVP_PKEY_set1_RSA(EVP_PKEY *pkey, RSA *key) {
   if (EVP_PKEY_assign_RSA(pkey, key)) {
     RSA_up_ref(key);
diff --git a/include/openssl/evp.h b/include/openssl/evp.h
index 378eb320f7c01df33850cca9d0111db32ffa6175..9eebf9c609abd31ed63c1f1c720c716d074e3f6d 100644
--- a/include/openssl/evp.h
+++ b/include/openssl/evp.h
@@ -156,6 +156,8 @@ OPENSSL_EXPORT int EVP_PKEY_type(int nid);
 // returned lower-level objects are considered to also mutate the |EVP_PKEY| and
 // may not be called concurrently with other operations on the |EVP_PKEY|.
 
+OPENSSL_EXPORT void *EVP_PKEY_get0(const EVP_PKEY *pkey);
+
 OPENSSL_EXPORT int EVP_PKEY_set1_RSA(EVP_PKEY *pkey, RSA *key);
 OPENSSL_EXPORT int EVP_PKEY_assign_RSA(EVP_PKEY *pkey, RSA *key);
 OPENSSL_EXPORT RSA *EVP_PKEY_get0_RSA(const EVP_PKEY *pkey);
